x-pack/plugins/task_manager/server/queries/task_claiming.ts

```diff
@@ -0,0 +1,484 @@
+        asTaskClaimEvent(
+          id,
+          asErr({ task: none, errorType: TaskClaimErrorType.CLAIMED_BY_ID_NOT_RETURNED })
+        )
+      ),
+    ]);
+
+    const stats = {
+      tasksUpdated,
+      tasksConflicted,
+      tasksRejected: documentsClaimedByIdOutOfCapacity.length,
+      tasksClaimed: documentsClaimedByIdWithinCapacity.length + documentsClaimedBySchedule.length,
+    };
+
+    if (docs.length !== stats.tasksClaimed + stats.tasksRejected) {
+      this.logger.warn(
+        `[Task Ownership error]: ${stats.tasksClaimed} tasks were claimed by Kibana, but ${
+          docs.length
+        } task(s) were fetched (${docs.map((doc) => doc.id).join(', ')})`
+      );
+    }
+
+    return {
+      stats,
+      docs: [...documentsClaimedByIdWithinCapacity, ...documentsClaimedBySchedule],
+    };
+  };
+
+  private async markAvailableTasksAsClaimed({
+    claimOwnershipUntil,
+    claimTasksById,
+    size,
+    taskTypes,
+  }: OwnershipClaimingOpts): Promise<UpdateByQueryResult> {
+    const { taskTypesToSkip = [], taskTypesToClaim = [] } = groupBy(
+      [...this.definitions],
+      ([type]) => (taskTypes.has(type) ? 'taskTypesToClaim' : 'taskTypesToSkip')
+    );
+    const taskMaxAttempts = taskTypesToClaim.reduce((accumulator, [type, { maxAttempts }]) => {
+      return { ...accumulator, [type]: maxAttempts || this.maxAttempts };
+    }, {});
+    const queryForScheduledTasks = mustBeAllOf(
+      // Either a task with idle status and runAt <= now or
+      // status running or claiming with a retryAt <= now.
+      shouldBeOneOf(IdleTaskWithExpiredRunAt, RunningOrClaimingTaskWithExpiredRetryAt)
+    );
+
+    // The documents should be sorted by runAt/retryAt, unless there are pinned
+    // tasks being queried, in which case we want to sort by score first, and then
+    // the runAt/retryAt.  That way we'll get the pinned tasks first.  Note that
+    // the score seems to favor newer documents rather than older documents, so
+    // if there are not pinned tasks being queried, we do NOT want to sort by score
+    // at all, just by runAt/retryAt.
+    const sort: SortOptions = [SortByRunAtAndRetryAt];
+    if (claimTasksById && claimTasksById.length) {
+      sort.unshift('_score');
+    }
+
+    const apmTrans = apm.startTransaction(`taskManager markAvailableTasksAsClaimed`, 'taskManager');
+    const result = await this.taskStore.updateByQuery(
+      asUpdateByQuery({
+        query: matchesClauses(
+          mustBeAllOf(
+            claimTasksById && claimTasksById.length
+              ? asPinnedQuery(claimTasksById, queryForScheduledTasks)
+              : queryForScheduledTasks
+          ),
+          filterDownBy(InactiveTasks)
+        ),
+        update: updateFieldsAndMarkAsFailed(
+          {
+            ownerId: this.taskStore.taskManagerId,
+            retryAt: claimOwnershipUntil,
+          },
+          claimTasksById || [],
+          [...taskTypes],
```

Comment:
Should this be `taskTypesToClaim` instead of all task types?

Reply:
It isn't all types :)
It's the types passed in as an argument, which _is_ `taskTypesToClaim `

That said, it clearly isn't clear enough ;) I'll rename it